Sources/Basic/TemporaryFile.swift

```diff
@@ -0,0 +1,172 @@
+        fileHandle = NSFileHandle(fileDescriptor: fd)
+    }
+
+    /// Convinence method for writing to the temporary file using OutputByteStream.
+    /// Note: This will write using the fileHandle and hence move the handle's filepointer.
+    ///
+    /// - Parameters:
+    ///     - write: Closure which provides the OutputByteStream object to write to.
+    ///
+    /// - Throws: SystemError
+    public func write(write: (OutputByteStream) -> Void) throws {
+        write(stream)
+        return try fputs(stream.bytes.contents, fileHandle)
+    }
+
+    /// Remove the temporary file before deallocating.
+    deinit { unlink(path) }
+
+    /// Create a default TemporaryFile object for duration of the closure passed to it.
+    ///
+    /// - Parameters:
+    ///     - body: Closure providing a default instance of TemporaryFile.
+    ///
+    /// - Throws: TempFileError
+    public static func withFile<T>(body: @noescape (TemporaryFile) throws -> T) throws -> T {
+        return try body(try TemporaryFile())
+    }
+
+    /// Determines the directory in which the temporary file should be created. Also makes
+    /// sure the returning path has a trailing forward slash.
+    ///
+    /// - Parameters:
+    ///     - dir: If present this will be the temporary directory.
+    ///
+    /// - Returns: Path to directory in which temporary file should be created.
+    private static func determineTempDirectory(_ dir: String? = nil) -> String {
+        // FIXME: Add other platform specific locations.
+        var tmpDir = dir ?? getenv("TMPDIR") ?? getenv("TEMP") ?? getenv("TMP") ?? "/tmp/"
+        if !tmpDir.hasSuffix("/") { tmpDir += "/" }
+        precondition(tmpDir.isDirectory)
+        return tmpDir
+    }
+
+    /// Read the temporary file contents as a UTF8 encoded string if possible.
+    /// This will use a new file handle to read the file and not touch the file handle
+    /// exposed by this class.
+    ///
+    /// - Throws: TempFileError.unicodeDecodingError
+    ///
+    /// - Returns: String containing contents of file.
+    public func readFileAsUTF8String() throws -> String {
+        // Open a new filehandle for reading the contents.
+        guard let fileHandle = NSFileHandle(forReadingAtPath: path) else {
+            fatalError("Unable to open file handle for temporary file at \(path)")
+        }
+        // Read till end of the file.
+        guard let contents = String(data: fileHandle.readDataToEndOfFile(), encoding: NSUTF8StringEncoding) else {
+            throw TempFileError.unicodeDecodingError 
+        }
+        return contents
+    }
+}
+
+extension TemporaryFile: CustomStringConvertible {
+    public var description: String {
+        return "{\n" +
+               "  TemporaryFile\n" +
+               "  prefix: \(prefix)\n" +
+               "  suffix: \(suffix)\n" +
+               "  path: \(path)\n" +
+               "  fd: \(fd)\n" +
+               "  stream: " + stream.bytes.asReadableString + "\n}"
+    }
```

Comment:
I have a strong preference that `description` methods default to a simple succinct description suitable for representation in a debugger. I don't think they should be used for "dump everything I might want to know about this class" purposes.